addons/mass_mailing/controllers/main.py

```diff
@@ -167,23 +168,15 @@ def view(self, mailing_id, email=None, res_id=None, token=""):
             if not self._valid_unsubscribe_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
                 raise exceptions.AccessDenied()
 
-            res = mailing.convert_links()
-            base_url = mailing.get_base_url().rstrip('/')
-            urls_to_replace = [
-                (base_url + '/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id)),
-                (base_url + '/view', mailing._get_view_url(email, res_id))
-            ]
-            for url_to_replace, new_url in urls_to_replace:
-                if url_to_replace in res[mailing_id]:
-                    res[mailing_id] = res[mailing_id].replace(url_to_replace, new_url if new_url else '#')
-
-            res[mailing_id] = res[mailing_id].replace(
-                'class="o_snippet_view_in_browser"',
-                'class="o_snippet_view_in_browser" style="display: none;"'
-            )
+            html = str(request.env['mail.render.mixin'].sudo()._render_template_qweb(
+                mailing.body_html, mailing.mailing_model_real, [res_id], add_context=None, options={})[res_id])
+            # Update generic URLs (without parameters) to final ones
+            html = html.replace('/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id))
```

Comment:
🔥 XSS

If you can control the base URL, you can have XSS here
```
http://thisisatest" onclick="alert(1)" qsd
```

To be honest it's "not really" an issue because only administrator can control it, but it expose an issue of manipulating HTML like this: quote are not escaped

Can you ensure `'"<>` are always escaped ? (or at least removed if it's too anoying)

Reply:
Yes indeed, it is better to be careful even if the URL doesn't come from the user.
I have ensured that the mentioned char are removed from the URL.